components/Controls.tsx

```diff
@@ -9,11 +9,29 @@ import { compareVersions } from 'compare-versions'
 
 export interface BikeControlsArgs {
     bike: Bike
-    api: Api
+    api: Api | undefined
     disconnect: () => void
 }
 
 export default function BikeControls({ bike, api, disconnect }: BikeControlsArgs) {
+    // Check if api is available
```

Comment:
> I would prefer this code: This way you don't have the same code twice
> 
> ```tsx
> return (
>         <BikeContext.Provider value={bike}>
>             <ApiContext.Provider value={api}>
>                 <BikeStats bike={bike} />
>                 <SpeedLimit bike={bike} />
>                 <PowerLevel bike={bike} />
>                 <BellTone bike={bike} />
>                 <SoundBoard />
>                 {api ?
>                     <>
>                         <ShareBike bike={bike} api={api} />
>                         <CurrentShares bike={bike} api={api} />
>                     </>
>                     : undefined}
>                 <Button onClick={disconnect} secondary>
>                     Disconnect bike
>                 </Button>
>             </ApiContext.Provider>
>         </BikeContext.Provider>
>     )
> ```

Actually that won't work because of the ApiContext provider

Reply:
What do you think about this?
```dart
    const bikeStats = (
        <>
            <BikeStats bike={bike} />
            <SpeedLimit bike={bike} />
            <PowerLevel bike={bike} />
            <BellTone bike={bike} />
            <SoundBoard />
            {api ?
                <>
                    <ShareBike bike={bike} api={api!} />
                    <CurrentShares bike={bike} api={api!} />
                </>
                : undefined}
            <Button onClick={disconnect} secondary>Disconnect bike</Button>
        </>
    )

    return (
        <BikeContext.Provider value={bike}>
            {api ? (
                <ApiContext.Provider value={api!}>
                    {bikeStats}
                </ApiContext.Provider>
            ) : (
                bikeStats
            )}
        </BikeContext.Provider>
    )

Reply:
Oh wait I see my example above didn't work :),
Maybe something like this?:

```tsx
return (
        <BikeContext.Provider value={bike}>
                <BikeStats bike={bike} />
                <SpeedLimit bike={bike} />
                <PowerLevel bike={bike} />
                <BellTone bike={bike} />
                <SoundBoard />
                {api ?
                    <ApiContext.Provider value={api}>
                        <ShareBike bike={bike} api={api} />
                        <CurrentShares bike={bike} api={api} />
                    </ApiContext.Provider >
                    : undefined}
                <Button onClick={disconnect} secondary>
                    Disconnect bike
                </Button>
        </BikeContext.Provider>
    )
```

Reply:
> Oh wait I see my example above didn't work :), Maybe something like this?:
> 
> ```tsx
> return (
>         <BikeContext.Provider value={bike}>
>                 <BikeStats bike={bike} />
>                 <SpeedLimit bike={bike} />
>                 <PowerLevel bike={bike} />
>                 <BellTone bike={bike} />
>                 <SoundBoard />
>                 {api ?
>                     <ApiContext.Provider value={api}>
>                         <ShareBike bike={bike} api={api} />
>                         <CurrentShares bike={bike} api={api} />
>                     </ApiContext.Provider >
>                     : undefined}
>                 <Button onClick={disconnect} secondary>
>                     Disconnect bike
>                 </Button>
>         </BikeContext.Provider>
>     )
> ```

I changed it into this now, will test if everything still functions later

Reply:
Works 👍🏻